haskell-api.md

```diff
@@ -260,40 +260,43 @@ Example:
         / StudentCoursesR GET   --> Handlers/V3/Students/Student/Courses.hs
 ```
 
-## Module Structure
+## API Types
 
-- Prefer in-module request/response types
+Prefer distinct types for requests and responses that are only used for that. We
+call these "API Types". Do not use persistent Entity types as API request or
+response types.
 
-  - Request type naming: `{Resource}{Method}Body`
-  - Response type naming: `{Resource}{Method}`
-  - NOTE: list responses are `[{Resource}{Method}]`, not `{Resources}{Method}`
+1. A response API Type should be named `Api{Resource}` and should be re-used
+   anywhere `Resource` appears in the API.
 
-  ```hs
-  -- Good
-  data TeacherGet
+   If the same concept takes on a different shape in different places, it should
+   be considered a different Resource, with a different name. For example,
+   `ApiTeacher`, `ApiAbbreviatedTeacher`, `ApiTeacherWithSchool`,
+   `ApiUsageReportTeacher`.
 
-  getTeachersR :: Handler [TeacherGet]
-  getTeachersR = undefined
+1. A request API Type should be named `Api{Action}{Resource}({Target})`.
```

Comment:
I always forget about the "it's okay to duplicate numbers in md bullets" thing but then I happily remember that md's got your back!